core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/CrtInputStream.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.crt.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.http.HttpRequestBodyStream;
+import software.amazon.awssdk.http.ContentStreamProvider;
+
+@SdkInternalApi
+public final class CrtInputStream implements HttpRequestBodyStream {
+    private final ContentStreamProvider provider;
+    private final int bufSize;
+    private final byte[] readBuffer;
+    private InputStream providerStream;
+
+    CrtInputStream(ContentStreamProvider provider, int bufSize) {
+        this.provider = provider;
+        this.bufSize = bufSize;
+        this.readBuffer = new byte[bufSize];
+    }
+
+    @Override
+    public boolean sendRequestBody(ByteBuffer bodyBytesOut) {
+        int read = 0;
+
+        try {
+            if (providerStream == null) {
+                createNewStream();
+            }
+
+            int toRead = Math.min(bufSize, bodyBytesOut.remaining());
+            read = providerStream.read(readBuffer, 0, toRead);
+
+            if (read > 0) {
+                bodyBytesOut.put(readBuffer, 0, read);
+            }
+        } catch (IOException ioe) {
+            throw new RuntimeException(ioe);
+        }
+
+        return read < 0;
+    }
+
+    @Override
+    public boolean resetPosition() {
+        if (provider == null) {
+            throw new IllegalStateException("Cannot reset position while provider is null");
+        }
```

Comment:
When will provider be null?

Reply:
Never, unless the object is constructed with null provider. IMO, we should validate the param when constructing (this is a copied class).